instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt

```diff
@@ -200,4 +212,75 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
     }
+
+    @Test
+    fun reroute_is_cancelled_when_the_user_returns_to_route() = sdkTest {
+        val responseDelay = 4000L
+        val mockRoute = RoutesProvider.dc_very_short(activity)
+        val originLocation = mockRoute.routeWaypoints.first()
+        val initialLocation = mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = originLocation.latitude()
+            longitude = originLocation.longitude()
+        }
+        val offRouteLocationUpdate = mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = originLocation.latitude() + 0.002
+            longitude = originLocation.longitude()
+        }
+
+        mockWebServerRule.requestHandlers.addAll(mockRoute.mockRequestHandlers)
+        val requestHandler = MockDirectionsRequestHandler(
+            profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+            jsonResponse = readRawFileText(activity, R.raw.reroute_response_dc_very_short),
+            expectedCoordinates = mockRoute.routeWaypoints,
+            relaxedExpectedCoordinates = false
+        )
+        val rerouteRequestHandler = MockDirectionsRequestHandler(
+            profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+            jsonResponse = readRawFileText(activity, R.raw.reroute_response_dc_very_short),
+            expectedCoordinates = listOf(
+                Point.fromLngLat(
+                    offRouteLocationUpdate.longitude,
+                    offRouteLocationUpdate.latitude
+                ),
+                mockRoute.routeWaypoints.last()
+            ),
+            relaxedExpectedCoordinates = false
+        )
+        rerouteRequestHandler.jsonResponseModifier = DelayedResponseModifier(responseDelay)
+        mockWebServerRule.requestHandlers.add(requestHandler)
+        mockWebServerRule.requestHandlers.add(rerouteRequestHandler)
+
+        val originalRoutes = mapboxNavigation.requestRoutes(
+            RouteOptions.builder()
+                .applyDefaultNavigationOptions()
+                .applyLanguageAndVoiceUnitOptions(activity)
+                .baseUrl(mockWebServerRule.baseUrl)
+                .coordinatesList(mockRoute.routeWaypoints)
+                .build()
+        ).getSuccessfulResultOrThrowException().routes
+        mapboxNavigation.startTripSession()
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(originalRoutes)
+
+        val rerouteStates = mutableListOf<RerouteState>()
+        val rerouteStatesObserver = RerouteController.RerouteStateObserver { rerouteState ->
+            rerouteStates.add(rerouteState)
+        }
+        mapboxNavigation.getRerouteController()!!.registerRerouteStateObserver(rerouteStatesObserver)
+
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+        mapboxNavigation.offRouteUpdates().filter { it }.first()
+        mockLocationReplayerRule.loopUpdate(initialLocation, times = 120)
+        mapboxNavigation.offRouteUpdates().filter { !it }.first()
```

Comment:
why is it needed 2 times? 
```
        mapboxNavigation.offRouteUpdates().filter { it }.first()
        mapboxNavigation.offRouteUpdates().filter { !it }.first()
```

Reply:
The second one waits for offRoute update when it's false, not true. I want to make sure we return to the route. Should I write `it == false` or use `filterNot` to avoid the confusion?

Reply:
now I see 👍 could you leave a comment on that here we're waiting and what is actually waiting for